lib/Conversion/PolyToStandard/PolyToStandard.cpp

```diff
@@ -72,44 +72,50 @@ struct ConvertAdd : public OpConversionPattern<AddOp> {
   // element-wise as elements of the ring, so they are performed modulo the
   // coefficient modulus.
   //
-  // Modular addition is performed by sign extending both arguments and
-  // performing (N+1)-bit addition. Then, if z = (x + y) is the (N+1)-bit sum,
-  // we return
-  //   truncate(z >= mod ? z - mod : z)
+  // To perform modular addition, assume that `cmod` is the coefficient modulus
+  // of the ring, and that `N` is the bitwidth used to store the ring elements.
+  // This may be much larger than `log_2(cmod)`.
+  //
+  // Let `x` and `y` be the inputs to modular addition, then:
+  //    c1, n1 = addui_extended(x, y)
+  // If the coefficient modulus divides `2^N`, then return
+  //    c0 = c1 % cmod
+  // Otherwise, compute the adjusted result:
+  //    c0 = ((c1 % cmod) + (n1 * 2^N % cmod)) % cmod
   LogicalResult matchAndRewrite(
       AddOp op, OpAdaptor adaptor,
       ConversionPatternRewriter &rewriter) const override {
     auto type = adaptor.getLhs().getType();
-    auto elementTy =
-        type.cast<RankedTensorType>().getElementType().cast<IntegerType>();
-    auto extType = IntegerType::get(getContext(), elementTy.getWidth() + 1,
-                                    elementTy.getSignedness());
-    auto extShapeType = type.cast<ShapedType>().clone(extType);
-
-    mlir::Value extLhsOp = rewriter.create<arith::ExtUIOp>(
-        op.getLoc(), extShapeType, adaptor.getLhs());
-    mlir::Value extRhsOp = rewriter.create<arith::ExtUIOp>(
-        op.getLoc(), extShapeType, adaptor.getRhs());
-
-    auto addOp = rewriter.create<arith::AddIOp>(op.getLoc(), extShapeType,
-                                                extLhsOp, extRhsOp);
-
-    APInt mod =
-        op.getType().cast<PolynomialType>().getRing().coefficientModulus();
-    assert(mod != 0 && "coefficient modulus must not be zero");
-    auto modConstOp = rewriter.create<arith::ConstantOp>(
-        op.getLoc(), DenseElementsAttr::get(extShapeType, {mod}));
-
-    auto geCmp = rewriter.create<arith::CmpIOp>(
-        addOp.getLoc(), arith::CmpIPredicate::uge, addOp, modConstOp);
-    auto selectOp = rewriter.create<arith::SelectOp>(
-        geCmp->getLoc(), geCmp,
-        rewriter.create<arith::SubIOp>(geCmp->getLoc(), addOp, modConstOp),
-        addOp);
+    APInt mod = op.getType().cast<PolyType>().getRing().coefficientModulus();
+    auto cmod = rewriter.create<arith::ConstantOp>(
+        op.getLoc(), DenseElementsAttr::get(cast<ShapedType>(type), {mod}));
```

Comment:
Since you're generating a bunch of sequential statements, consider using ImplicitLocOpBuilder https://github.com/llvm/llvm-project/blob/d671126ad097ffd1bb19322005902676945862c2/mlir/include/mlir/IR/ImplicitLocOpBuilder.h#L23